libc-bottom-half/crt/crt1-command.c

```diff
@@ -3,10 +3,63 @@
 extern void __wasi_init_tp(void);
 #endif
 #include <wasi/api.h>
+#include <stdlib.h>
+#include <sysexits.h>
 extern void __wasm_call_ctors(void);
 extern int __main_void(void);
 extern void __wasm_call_dtors(void);
 
+// The user's `main` function, expecting arguments.
+int __main_argc_argv(int argc, char *argv[]);
+
+// If the user's `main` function expects arguments, the compiler will rename
+// it to `__main_argc_argv`, and this version will get linked in, which
+// initializes the argument data and calls `__main_argc_argv`.
+__attribute__((__weak__, nodebug))
+int __main_void(void) {
+    __wasi_errno_t err;
+
+    // Get the sizes of the arrays we'll have to create to copy in the args.
+    size_t argv_buf_size;
+    size_t argc;
+    err = __wasi_args_sizes_get(&argc, &argv_buf_size);
+    if (err != __WASI_ERRNO_SUCCESS) {
+        _Exit(EX_OSERR);
+    }
+
+    // Add 1 for the NULL pointer to mark the end, and check for overflow.
+    size_t num_ptrs = argc + 1;
+    if (num_ptrs == 0) {
+        _Exit(EX_SOFTWARE);
+    }
+
+    // Allocate memory for storing the argument chars.
+    char *argv_buf = malloc(argv_buf_size);
```

Comment:
does this mean to rely on `--gc-sections`? i'm not sure if it's a good idea.

Reply:
Sorry, I'm not sure I understand what you mean.  My understanding is that a command-style app will need to define a either `main` function that takes no arguments or one that takes `argc` and `argv` arguments.  If the former, this `__main_void` will not be linked in at all (since it's a weak symbol).  If the latter, this function will (and should) be used.  Under what circumstances would `--gc-sections` affect the result?

Reply:
I'm not sure I understand either, but regardless, `--gc-sections` is the default in wasm-ld.  

Furthermore, `-ffunctions-sections` and `-fdata-sections` are also enabled by default in the llvm backend, and actually there is no way to disable `-ffunctions-sections`.

Reply:
i meant it could break `--no-gc-sections`.

Reply:
@yamt Can you provide an example of a case where this would break `--no-gc-sections`?  I'm still having trouble seeing the problem.

Reply:
I'm also curious if you have a use case for `-Wl,--no-gc-sections`.  I'm not saying we should keep it working, but I wonder if you have a use case in mind for it.  I'm struggling to think of one.

Reply:
> @yamt Can you provide an example of a case where this would break `--no-gc-sections`? I'm still having trouble seeing the problem.

just link a module with `--no-gc-sections`?
```
spacetanuki% cat a.c
int main() {return 0;}
spacetanuki% /opt/wasi-sdk/bin/clang --sysroot=${HOME}/git/wasi-libc/sysroot -Xlinker --no-gc-sections a.c
wasm-ld: error: /Users/yamamoto/git/wasi-libc/sysroot/lib/wasm32-wasi/crt1-command.o: undefined symbol: main
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
spacetanuki%
```

Reply:
> I'm also curious if you have a use case for `-Wl,--no-gc-sections`. I'm not saying we should keep it working, but I wonder if you have a use case in mind for it. I'm struggling to think of one.

i don't have concrete use cases either.
but i can imagine there can be "creative" users. like embedding a magic signature in a module, which is otherwise unused. especially for apps which are ported from non-wasm environment.

Reply:
> just link a module with `--no-gc-sections`?

Thanks.  I still don't understand _why_ `--no-gc-sections` makes a difference here, but clearly it does.

Reply:
Can you explain @yamt ?  What made you suspect this would be an issue with `--no-gc-sections`?   Or as it simply via experimentation that you discovered this issue?  

Normally `--no-gc-sections` should not effect whether a linker succeeds or not.. so this could be rare case.. or a linker bug.   Nothing jumps out at me here as to why this would cause that undefined symbol error, but I have yet to try to repro this locally.

Reply:
Confirmed.. this is behaviour also occurs with GNU ld, so we can't have this symbol live in the same compilation unit as `_start`

Reply:
Maybe we can put this symbol into compiler-rt, which I assume will remain statically linked.   For gcc it looks like the compiler rt is split into a shared part (libgcc_s) and a non-shared part (libgcc)

Reply:
Moving it to compiler-rt would require an LLVM change, right?  If so, I'm assuming it's too late to get that into LLVM 17, and I'd prefer not to wait another 6 months to get everything aligned.  Perhaps we should just move it back to main_void.c (where it was originally) for the time being?  Alternatively, we could patch LLVM in wasi-sdk while we're waiting for LLVM 18.

Reply:
In wasi-sdk we can choose what we package into our compiler-rt library.. I don't see why we can't add to it.

Either that or we could create some other object that clang can link in (e.g. crt0.o or crtend.o).. but that would also require a clang driver change.

Reply:
@sbc100 Please check my understanding of your first suggestion above:

1. We build `libclang_rt.builtins-wasm32.a` as usual in `wasi-sdk`
2. We also build `__main_void.o` from `__main_void.c` (which would contain the `__main_void` definition we've been discussing).
3. We add `__main_void.o` to `libclang_rt.builtins-wasm32.a`

Does that sound right?  If so, any objection to putting `__main_void.c` back where it was in `libc-bottom-half/sources` but omitting it from libc, using it only as described above?

Reply:
That is certainly an option... I can see an argument though for wanting to keep libclang_rt.builtins-wasm32.a pure.  Are there any other examples of files that we don't want to in the shared library but make more sense linking statically?

Reply:
> That is certainly an option... I can see an argument though for wanting to keep libclang_rt.builtins-wasm32.a pure. Are there any other examples of files that we don't want to in the shared library but make more sense linking statically?

Nothing else has come up to my knowledge.

BTW, having `__main_void` in the shared library is not really a _problem_ -- just slightly awkward since it won't be used for reactors.  As long as `__main_argc_argv` is declared weak (which is what started this whole discussion), there's no burden left to the application developer.  So we always have an option of going back to where we started and adding a comment explaining the situation per https://github.com/WebAssembly/wasi-libc/pull/429#discussion_r1290434731.

Reply:
i guess that putting back __main_void to libc involves the least effort for now.

Reply:
@sbc100 How would you feel if I put `__main_void` back in libc for the time being?  I can add a comment about why the `__main_argc_argv` declaration is weak and that our long-term goal is to move `__main_void` to a crt*.o file that is linked in as appropriate in a `--no-gc-sections`-friendly way.

Reply:
Sure, that sounds fine, given that moving proved problematic.